### PR TITLE
Changing TokenService MSI sidecar specialization API to 2022-01-01

### PR DESCRIPTION
**Pull request by karshinlin**
### Issue describing the changes in this PR

Adding new specialization API in Token Service (MSI sidecar) to support delegated identities.

This builds off a change in the App Service platform that adds a separates SystemAssigned, UserAssigned, and Delegated identities.

### Pull request checklist

* [X] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [X] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [ ] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [X] My changes **do not** require diagnostic events changes
    * Otherwise: I have added/updated all related diagnostic events and their documentation (Documentation issue linked to PR)
* [X] I have added all required tests (Unit tests, E2E tests)

### Additional information

Additional PR information

### CODE REVIEW

**Comment by pragnagopa** [comment]
Marking this PR as draft until new TokenService V2 - that has new APIs is rolled out to Production

**Review by yojagad — APPROVED** [review]

File: src/WebJobs.Script.WebHost/Models/MSIContext.cs [hunk 1]
```diff
@@ -12,5 +12,11 @@ public class MSIContext
         public string MSISecret { get; set; }
 
         public IEnumerable<ManagedServiceIdentity> Identities { get; set; }
+
+        public ManagedServiceIdentity SystemAssignedIdentity { get; set; }
+
+        public IEnumerable<ManagedServiceIdentity> UserAssignedIdentities { get; set; }
```
Comment (pragnagopa) [inline src/WebJobs.Script.WebHost/Models/MSIContext.cs:18]: please see if there are existing unit tests. if yes, expand to include tests for new properties

Comment (karshinlin) [reply src/WebJobs.Script.WebHost/Models/MSIContext.cs:18]: There are existing tests here: https://github.com/Azure/azure-functions-host/blob/dev/test/WebJobs.Script.Tests.Integration/Management/InstanceManagerTests.cs#L571-L571

For the most part, MSI sidecar initialization is just a passthrough from perspective of Functions Host (serialize and forward payload). I think this should be by design to reduce added complexity of depending on MSIContext within Functions Host.

Test scenarios already covered:
1. If MSI endpoint doesn't exist
2. If MSIContext from DataRole is null
3. MSI sidecar needs no warmup ping
4. Successful specialization

File: src/WebJobs.Script/ScriptConstants.cs [hunk 2]
```diff
@@ -166,7 +166,7 @@ public static class ScriptConstants
         public const string LinuxMetricEventStreamName = "MS_FUNCTION_METRICS";
         public const string LinuxFunctionDetailsEventStreamName = "MS_FUNCTION_DETAILS";
         public const string LinuxAzureMonitorEventStreamName = "MS_FUNCTION_AZURE_MONITOR_EVENT";
-        public const string LinuxMSISpecializationStem = "/api/specialize?api-version=2017-09-01";
```
Comment (balag0) [inline src/WebJobs.Script/ScriptConstants.cs:169]: I think you might have mentioned this elsewhere. but just to confirm :)
1)we will only release this change after new token svc image is already released fully
2)the new token svc image supports the old api version 2017-09-01.

Correct?

Comment (karshinlin) [reply src/WebJobs.Script/ScriptConstants.cs:169]: Yes for both. Finished new token service release yesterday 3/16 across all of public.

File: src/WebJobs.Script.WebHost/Models/MSIContext.cs (hunk 1 above)
Comment (karshinlin) [reply src/WebJobs.Script.WebHost/Models/MSIContext.cs:18]: Added an additional test for properties that are expected during msi specialization. @pragnagopa

**Review by pragnagopa — APPROVED** [review]

**Pull request merged**